File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        if (n <= k) {
            return "0";
        }
        if (k == 0) {
            return num;
        }
        stack<char> st;
        st.push(num[0]);
        for (int i = 1; i < n; ++i) {
            while (k > 0 and !st.empty() and st.top() > num[i]) {
                st.pop();
                --k;
            }
            st.push(num[i]);
            if (st.size() == 1 and num[i] == '0') {
                st.pop();
            }
        }
        while (k > 0 and !st.empty()) {
            st.pop();
            --k;
        }
        string res;
        while (!st.empty()) {
            res.push_back(st.top());
            st.pop();
        }
        if (res.size() == 0) {
            return "0";
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        
        for d in num:
            while stack and stack[-1] > d and k > 0:
                stack.pop()
                k -= 1
            stack.append(d)
            
        while stack and k > 0:
            stack.pop()
            k -= 1
            
        i = 0
        
        while i < len(stack) and stack[i] == '0':
            i += 1
            
        res = "".join(stack[i:])
        
        return res or "0"